File: test-suite/src/test/java/io/micronaut/gcp/pubsub/subscriber/ContentTypeSubscriberTest.java
package io.micronaut.gcp.pubsub.subscriber;

import com.google.pubsub.v1.PubsubMessage;
import io.micronaut.context.annotation.Property;
import io.micronaut.context.annotation.Requires;
import io.micronaut.context.event.BeanCreatedEvent;
import io.micronaut.context.event.BeanCreatedEventListener;
import io.micronaut.core.annotation.NonNull;
import io.micronaut.gcp.pubsub.annotation.PubSubClient;
import io.micronaut.gcp.pubsub.annotation.Topic;
import io.micronaut.gcp.pubsub.support.Animal;
import io.micronaut.http.MediaType;
import io.micronaut.test.extensions.junit5.annotation.MicronautTest;
import jakarta.inject.Inject;
import jakarta.inject.Singleton;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.mockito.ArgumentCaptor;
import org.mockito.Mockito;

import java.nio.charset.StandardCharsets;

import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.spy;
import static org.mockito.Mockito.timeout;
import static org.mockito.Mockito.verify;

@MicronautTest
@Property(name = "spec.name", value = "ContentTypeSubscriberTest")
public class ContentTypeSubscriberTest {

    @Inject
    TestPublisher publisher;

    @Inject
    ContentTypeSubscriber subscriber;

    @BeforeEach
    void setup() {
        Mockito.reset(subscriber);
    }

    @Test
    void testRawBytes() {
        byte[] bytesSent = "foo".getBytes(StandardCharsets.UTF_8);
        publisher.publishRaw(bytesSent);

        ArgumentCaptor<byte[]> bytesReceived = ArgumentCaptor.forClass(byte[].class);
        verify(subscriber, timeout(3000)).receiveRaw(bytesReceived.capture(), any());
        String decodedMessage = new String(bytesReceived.getValue(), StandardCharsets.UTF_8);
        Assertions.assertEquals("foo", decodedMessage);
    }

    @Test
    void testNativeMessage() {
        byte[] bytesSent = "foo".getBytes(StandardCharsets.UTF_8);
        publisher.publishNative(bytesSent);

        ArgumentCaptor<PubsubMessage> messageReceived = ArgumentCaptor.forClass(PubsubMessage.class);
        verify(subscriber, timeout(3000)).receiveNative(messageReceived.capture());
        String decodedMessage = messageReceived.getValue().getData().toString(StandardCharsets.UTF_8);
        Assertions.assertEquals("foo", decodedMessage);
    }

    @Test
    void testJsonPojo() {
        Animal dog = new Animal("dog");
        publisher.publishAnimal(dog);

        ArgumentCaptor<Animal> messageReceived = ArgumentCaptor.forClass(Animal.class);
        verify(subscriber, timeout(3000)).receivePojo(messageReceived.capture(), any());
        Assertions.assertNotNull(messageReceived.getValue());
        Assertions.assertEquals("dog", messageReceived.getValue().getName());
    }

    @Test
    void testXmlPojo() {
        Animal dog = new Animal("cat");
        publisher.publishAnimalAsXml(dog);

        ArgumentCaptor<Animal> messageReceived = ArgumentCaptor.forClass(Animal.class);
        verify(subscriber, timeout(3000)).receiveXML(messageReceived.capture(), any());
        Assertions.assertNotNull(messageReceived.getValue());
        Assertions.assertEquals("cat", messageReceived.getValue().getName());
    }

    @Singleton
    @PubSubClient
    @Requires(property = "spec.name", value = "ContentTypeSubscriberTest")
    interface TestPublisher {
        @Topic("raw-subscription") void publishRaw(byte[] payload);
        @Topic("native-subscription") void publishNative(byte[] payload);
        @Topic("animals") void publishAnimal(Animal animal);
        @Topic(value = "animals-legacy", contentType = MediaType.APPLICATION_XML) void publishAnimalAsXml(Animal animal);
    }

    @Singleton
    @Requires(property = "spec.name", value = "ContentTypeSubscriberTest")
    static class SubscriberCreatedListener implements BeanCreatedEventListener<ContentTypeSubscriber> {
        @Override
        public ContentTypeSubscriber onCreated(@NonNull BeanCreatedEvent<ContentTypeSubscriber> event) {
            return spy(event.getBean());
        }
    }
}

File: test-suite-python/src/test/python/micronaut/gcp/pubsub/subscriber/ContentTypeSubscriberTest.py
from abc import ABC, abstractmethod
from typing import Annotated

import java
from com.google.pubsub.v1 import PubsubMessage
from jakarta.inject import Inject, Singleton
from java.util.concurrent import TimeUnit
from micronaut.context.annotation import Property, Requires
from micronaut.gcp.pubsub.annotation import PubSubClient, Topic
from micronaut.gcp.pubsub.support.Animal import Animal
from micronaut.test.extensions.junit5.annotation import MicronautTest
from org.junit.jupiter.api import BeforeEach, Test

from .ContentTypeMockMessageProcessor import ContentTypeMockMessageProcessor
from micronaut.gcp.pubsub.subscriber.MessageProcessor import MessageProcessor


@PubSubClient
@Requires(property="spec.name", value="ContentTypeSubscriberTest")
class ContentTypeTestPublisher(ABC):

    @Topic("raw-subscription")
    @abstractmethod
    def publish_raw(self, payload: bytes) -> None:
        ...

    @Topic("native-subscription")
    @abstractmethod
    def publish_native(self, payload: bytes) -> None:
        ...

    @Topic("animals")
    @abstractmethod
    def publish_animal(self, animal: Animal) -> None:
        ...

    @Topic(value="animals-legacy", contentType="application/xml")
    @abstractmethod
    def publish_animal_as_xml(self, animal: Animal) -> None:
        ...


@Property(name="spec.name", value="ContentTypeSubscriberTest")
@MicronautTest
class ContentTypeSubscriberTest:
    publisher: Annotated[ContentTypeTestPublisher, Inject]
    messageProcessor: Annotated[MessageProcessor, Inject]

    @BeforeEach
    def setup(self):
        self.messageProcessor.received = None

    def await_received(self):
        for _ in range(60):
            if self.messageProcessor.received is not None:
                return self.messageProcessor.received
            TimeUnit.MILLISECONDS.sleep(50)
        assert self.messageProcessor.received is not None, "no message received"

    @Test
    def test_raw_bytes(self):
        self.publisher.publish_raw("foo".encode())

        received = self.await_received()
        assert bytes(received).decode() == "foo"

    @Test
    def test_native_message(self):
        self.publisher.publish_native("foo".encode())

        received = self.await_received()
        assert java.instanceof(received, PubsubMessage)
        assert received.getData().toStringUtf8() == "foo"

    @Test
    def test_json_pojo(self):
        self.publisher.publish_animal(Animal("dog"))

        received = self.await_received()
        assert received.name == "dog"

    @Test
    def test_xml_pojo(self):
        self.publisher.publish_animal_as_xml(Animal("cat"))

        received = self.await_received()
        assert received.name == "cat"
